crates/holochain/src/conductor/manager/mod.rs

```diff
@@ -332,128 +248,209 @@ pub fn handle_shutdown(result: Result<TaskManagerResult, tokio::task::JoinError>
     }
 }
 
+/// Each task has a group, and here they are
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum TaskGroup {
+    /// Tasks which are associated with the conductor as a whole
+    Conductor,
+    /// Tasks which are associated with a particular DNA space
+    Dna(Arc<DnaHash>),
+    /// Tasks which are associated with a particular running Cell
+    Cell(CellId),
+}
+
+/// Channel sender for task outcomes
+pub type OutcomeSender = futures::channel::mpsc::Sender<(TaskGroup, TaskOutcome)>;
+/// Channel receiver for task outcomes
+pub type OutcomeReceiver = futures::channel::mpsc::Receiver<(TaskGroup, TaskOutcome)>;
+
 /// A collection of channels and handles used by the Conductor to talk to the
 /// TaskManager task
+#[derive(Clone)]
 pub struct TaskManagerClient {
-    /// Channel on which to send info about tasks we want to manage
-    task_add_sender: mpsc::Sender<ManagedTaskAdd>,
-
-    /// Sending a message on this channel will broadcast to all managed tasks,
-    /// telling them to shut down
-    task_stop_broadcaster: StopBroadcaster,
-
-    /// The main task join handle to await on.
-    /// The conductor is intended to live as long as this task does.
-    /// It can be moved out, hence the Option. If this is None, then the
-    /// handle was already moved out.
-    run_handle: Option<TaskManagerRunHandle>,
+    tm: Arc<Mutex<Option<task_motel::TaskManager<TaskGroup, TaskOutcome>>>>,
 }
 
 impl TaskManagerClient {
-    /// Constructor
-    pub fn new(
-        task_add_sender: mpsc::Sender<ManagedTaskAdd>,
-        task_stop_broadcaster: StopBroadcaster,
-        run_handle: TaskManagerRunHandle,
-    ) -> Self {
+    /// Construct the TaskManager and the outcome channel receiver
+    pub fn new(tx: OutcomeSender) -> Self {
+        let tm = task_motel::TaskManager::new(tx, |g| match g {
+            TaskGroup::Conductor => None,
+            TaskGroup::Dna(_) => Some(TaskGroup::Conductor),
+            TaskGroup::Cell(cell_id) => Some(TaskGroup::Dna(Arc::new(cell_id.dna_hash().clone()))),
+        });
         Self {
-            task_add_sender,
-            task_stop_broadcaster,
-            run_handle: Some(run_handle),
+            tm: Arc::new(Mutex::new(Some(tm))),
         }
     }
 
-    /// Accessor
-    pub fn task_add_sender(&self) -> &mpsc::Sender<ManagedTaskAdd> {
-        &self.task_add_sender
+    /// Stop all tasks and await their completion.
+    pub fn stop_all_tasks(&self) -> ShutdownHandle {
+        if let Some(tm) = self.tm.lock().as_mut() {
+            tokio::spawn(tm.stop_group(&TaskGroup::Conductor))
+        } else {
+            tracing::warn!("Tried to shutdown task manager while it's already shutting down");
+            tokio::spawn(async move {})
+        }
+    }
```

Comment:
Are you needing to abort these anywhere? If not, you could just make these functions `async` and not have to do that awkward empty `tokio::spawn(async move {})`.

Reply:
(or return `impl Future<Output = ()> + 'static + Send`) if you need to make sure they don't hold the `self` reference...

Reply:
@neonphog it's not that I need to abort them, but I wanted to ensure that the `stop_group` future polls to completion, i.e. so you can either fire-and-forget, or await. But maybe that's overkill since there is only one place this is used, and it's awaited. And I guess we'll get compiler warnings if the future is unused, so the user would know to spawn a task. OK, I'll just make it async.